### ISSUE TITLE
Unable to specify the GPU device for training; it is set to GPU2 but GPU0 is being used.

### ISSUE THREAD

**Issue by Zarato2122** [huggingface/diffusers#10191]
### Describe the bug

I have already set the GPU for training to GPU2 in the program using the statement os.environ["CUDA_VISIBLE_DEVICES"] = "2". However, after starting the training, an error message indicates that GPU0 is being used, and the memory of my GPU0 is already insufficient.

Traceback (most recent call last):
  File "train_sdxl.py", line 1996, in <module>
    main(args)
  File "train_sdxl.py", line 1705, in main
    model_pred = unet(
  File "/home/ubuntu/miniconda3/envs/czc_SDXL/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1553, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/ubuntu/miniconda3/envs/czc_SDXL/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1562, in _call_impl
    return forward_call(*args, **kwargs)
  File "/data/ziqiang/czc/diffusers/src/diffusers/models/unets/unet_2d_condition.py", line 1281, in forward
    sample = upsample_block(
  File "/home/ubuntu/miniconda3/envs/czc_SDXL/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1553, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/ubuntu/miniconda3/envs/czc_SDXL/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1562, in _call_impl
    return forward_call(*args, **kwargs)
  File "/data/ziqiang/czc/diffusers/src/diffusers/models/unets/unet_2d_blocks.py", line 2551, in forward
    hidden_states = attn(
  File "/home/ubuntu/miniconda3/envs/czc_SDXL/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1553, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/ubuntu/miniconda3/envs/czc_SDXL/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1562, in _call_impl
    return forward_call(*args, **kwargs)
  File "/data/ziqiang/czc/diffusers/src/diffusers/models/transformers/transformer_2d.py", line 442, in forward
    hidden_states = block(
  File "/home/ubuntu/miniconda3/envs/czc_SDXL/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1553, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/ubuntu/miniconda3/envs/czc_SDXL/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1562, in _call_impl
    return forward_call(*args, **kwargs)
  File "/data/ziqiang/czc/diffusers/src/diffusers/models/attention.py", line 571, in forward
    ff_output = self.ff(norm_hidden_states)
  File "/home/ubuntu/miniconda3/envs/czc_SDXL/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1553, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/ubuntu/miniconda3/envs/czc_SDXL/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1562, in _call_impl
    return forward_call(*args, **kwargs)
  File "/data/ziqiang/czc/diffusers/src/diffusers/models/attention.py", line 1242, in forward
    hidden_states = module(hidden_states)
  File "/home/ubuntu/miniconda3/envs/czc_SDXL/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1553, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/ubuntu/miniconda3/envs/czc_SDXL/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1562, in _call_impl
    return forward_call(*args, **kwargs)
  File "/data/ziqiang/czc/diffusers/src/diffusers/models/activations.py", line 117, in forward
    hidden_states = self.proj(hidden_states)
  File "/home/ubuntu/miniconda3/envs/czc_SDXL/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1553, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/ubuntu/miniconda3/envs/czc_SDXL/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1562, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/ubuntu/miniconda3/envs/czc_SDXL/lib/python3.8/site-packages/torch/nn/modules/linear.py", line 117, in forward
    return F.linear(input, self.weight, self.bias)
torch.OutOfMemoryError: CUDA out of memory. Tried to allocate 320.00 MiB. GPU 0 has a total capacity of 47.53 GiB of which 169.31 MiB is free. Including non-PyTorch memory, this process has 47.35 GiB memory in use. Of the allocated memory 46.74 GiB is allocated by PyTorch, and 307.15 MiB is reserved by PyTorch but unallocated. If reserved but unallocated memory is large try setting PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True to avoid fragmentation.  See documentation for Memory Management  (https://pytorch.org/docs/stable/notes/cuda.html#environment-variables)

The above is the error information.

### Reproduction

if __name__ == "__main__":
    os.environ["CUDA_VISIBLE_DEVICES"] = "2"
    args = parse_args()
    main(args)

At the end of the program, I added the statement above to set the GPU.

Here is my complete code: [View Gist](https://gist.github.com/Zarato2122/ed6018c373e3701553b0074f4b9c4a95).
My code is primarily based on the lora-sdxl version, with only a few necessary parameters being modified for uploading.

Moreover，I start the training using python train_sdxl.py without utilizing an accelerator.

### System Info

- 🤗 Diffusers version: 0.32.0.dev0
- Platform: Linux-5.4.0-198-generic-x86_64-with-glibc2.17
- Running on Google Colab?: No
- Python version: 3.8.20
- PyTorch version (GPU?): 2.4.1+cu121 (True)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Huggingface_hub version: 0.26.3
- Transformers version: 4.46.3
- Accelerate version: 1.0.1
- PEFT version: 0.7.0
- Bitsandbytes version: not installed
- Safetensors version: 0.4.5
- xFormers version: not installed
- Accelerator: NVIDIA RTX A6000, 49140 MiB
NVIDIA RTX A6000, 49140 MiB
NVIDIA RTX A6000, 49140 MiB
- Using GPU in script?: <fill in>
- Using distributed or parallel set-up in script?: <fill in>

### Who can help?

@sayakpaul

**Comment by hlky** [comment]
`os.environ["CUDA_VISIBLE_DEVICES"] = "2"` may need to be set before other imports. You can also launch with `CUDA_VISIBLE_DEVICES="2" ...` e.g. `CUDA_VISIBLE_DEVICES="2" accelerate launch {script_name.py}`

**Comment by sayakpaul** [comment]
Yeah, `CUDA_VISIBLE_DEVICES=2 accelerate launch ...` is the safest bet and is what I personally follow too.

**Comment by Zarato2122** [comment]
import os
import torch

if torch.cuda.is_available():
    print("CUDA is available.")

    num_gpus = torch.cuda.device_count()
    print(f"Number of available GPUs: {num_gpus}")

    for i in range(num_gpus):
        print(f"GPU {i}: {torch.cuda.get_device_name(i)}")

    cuda_visible_devices = os.getenv('CUDA_VISIBLE_DEVICES')
    if cuda_visible_devices:
        print(f"CUDA_VISIBLE_DEVICES: {cuda_visible_devices}")
    else:
        print("CUDA_VISIBLE_DEVICES is not set.")
else:
    print("CUDA is not available.")


(czc_SDXL) ubuntu@ubuntu:/data/ziqiang/czc/diffusers/examples/dreambooth$ python train_sdxl.py
CUDA is available.
Number of available GPUs: 1
GPU 0: NVIDIA RTX A6000
CUDA_VISIBLE_DEVICES: 2

I tried to start the training using  CUDA_VISIBLE_DEVICES="2", but surprisingly, GPU0 was still being used. Then, I added statements at the beginning of the program to check the GPU devices. To my confusion, only GPU0 was visible. I clearly have three devices: GPU0, 1, and 2. However, it seems that only GPU0 is visible.

**Comment by Zarato2122** [comment]
<img width="501" alt="屏幕截图 2024-12-12 151521" src="https://github.com/user-attachments/assets/eeab8f8b-c18e-4734-8726-c632812da3b7" />
nvidia-smi

**Comment by Zarato2122** [comment]
<img width="526" alt="屏幕截图 2024-12-12 151626" src="https://github.com/user-attachments/assets/16e1b967-3519-46fe-af36-84ad34b71787" />

nvidia-smi

**Comment by sayakpaul** [comment]
What if you run the following?

```py
# name it `my_script.py`
import torch 

print(torch.cuda.available_devices())
```

```bash
CUDA_VISIBLE_DEVICES=0 python my_script.py
```

**Comment by Zarato2122** [comment]
(czc_SDXL) ubuntu@ubuntu:/data/ziqiang/czc/diffusers/examples/dreambooth$ CUDA_VISIBLE_DEVICES=0 python my_script.py
Traceback (most recent call last):
  File "my_script.py", line 4, in <module>
    print(torch.cuda.available_devices())
AttributeError: module 'torch.cuda' has no attribute 'available_devices'
Maybe this function is not supported.

I switched to the following program, which returned a result of 1.

# name it `my_script.py`
import torch 

print(torch.cuda.device_count())

(czc_SDXL) ubuntu@ubuntu:/data/ziqiang/czc/diffusers/examples/dreambooth$ CUDA_VISIBLE_DEVICES=0 python my_script.py
1

**Comment by sayakpaul** [comment]
Yeah sorry, should have been `device_count()`. It shows that it's working as expected i.e., PyTorch can only access a single GPU and so on. So, what's the current problem now?

**Comment by hlky** [comment]
Isn't this just how CUDA_VISIBLE_DEVICES works? You set CUDA_VISIBLE_DEVICES=2 then GPU 2 is referred to as GPU 0

**Comment by Zarato2122** [comment]
(czc_SDXL) ubuntu@ubuntu:/data/ziqiang/czc/diffusers/examples/dreambooth$ CUDA_VISIBLE_DEVICES=1 python train_sdxl.py02<?, ?it/s]
CUDA is available.
Number of available GPUs: 1
GPU 0: NVIDIA RTX A6000
CUDA_VISIBLE_DEVICES: 1

Sorry, I may not have expressed myself clearly. In fact, no matter if I set CUDA_VISIBLE_DEVICES to 1, 2, or 0, only one GPU is visible, which is GPU0, and it throws an out-of-memory error. If it were GPU2 that was being accessed, there shouldn't be an issue with insufficient memory.

**Comment by Zarato2122** [comment]
I have tried various commands, but the program always attempts to use GPU0 for training. I cannot specify the program to use GPU2 for training.

**Comment by hlky** [comment]
```
root@1d3e20184cab:/workspace# nvidia-smi 
Fri Dec 13 08:50:45 2024       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 565.57.01              Driver Version: 565.57.01      CUDA Version: 12.7     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA A40                     On  |   00000000:4F:00.0 Off |                    0 |
|  0%   24C    P8             22W /  300W |       1MiB /  46068MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
|   1  NVIDIA A40                     On  |   00000000:52:00.0 Off |                    0 |
|  0%   24C    P8             21W /  300W |       1MiB /  46068MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
                                                                                         
+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI        PID   Type   Process name                              GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|  No running processes found                                                             |
+-----------------------------------------------------------------------------------------+
```
```
root@1d3e20184cab:/workspace# CUDA_VISIBLE_DEVICES=1 python
Python 3.11.10 (main, Sep  7 2024, 18:35:41) [GCC 11.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> t = torch.randn([64, 64, 64, 64]).cuda()
>>> 
```
```
root@1d3e20184cab:/workspace# nvidia-smi 
Fri Dec 13 08:52:12 2024       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 565.57.01              Driver Version: 565.57.01      CUDA Version: 12.7     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA A40                     On  |   00000000:4F:00.0 Off |                    0 |
|  0%   22C    P8             21W /  300W |       1MiB /  46068MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
|   1  NVIDIA A40                     On  |   00000000:52:00.0 Off |                    0 |
|  0%   25C    P0             70W /  300W |     333MiB /  46068MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
                                                                                         
+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI        PID   Type   Process name                              GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
+-----------------------------------------------------------------------------------------+
```

`CUDA_VISIBLE_DEVICES=1` sets the device(s) visible to PyTorch etc. PyTorch sees device with index `1` as the only available device, therefore the index is `0`, it is still using the actual device index `1`.

It's using the correct GPU then OOM due to the script/arguments.

**Comment by Zarato2122** [comment]
example：
(czc_SDXL) ubuntu@ubuntu:/data/ziqiang/czc/diffusers/examples/dreambooth$ CUDA_VISIBLE_DEVICES=1,2 python train_sdxl.py
CUDA is available.
Number of available GPUs: 2
GPU 0: NVIDIA RTX A6000
GPU 1: NVIDIA RTX A6000
CUDA_VISIBLE_DEVICES: 1,2

torch.OutOfMemoryError: CUDA out of memory. Tried to allocate 320.00 MiB. GPU 0 has a total capacity of 47.53 GiB of which 156.69 MiB is free. Including non-PyTorch memory, this process has 47.35 GiB memory in use. Of the allocated memory 46.74 GiB is allocated by PyTorch, and 307.15 MiB is reserved by PyTorch but unallocated. If reserved but unallocated memory is large try setting PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True to avoid fragmentation.  See documentation for Memory Management  (https://pytorch.org/docs/stable/notes/cuda.html#environment-variables)



I understand that after setting up, the GPU IDs are reordered according to the sequence in the set, and PyTorch can indeed correctly locate the GPUs I have set. However, every time I start training, it reports an error indicating insufficient GPU memory. This happens regardless of whether I set CUDA_VISIBLE_DEVICES to 0,1,2 or 1,2 or just 2, etc. Moreover, the error message always points to insufficient memory on GPU0.

Regardless of how I set CUDA_VISIBLE_DEVICES, whether it's a single GPU, multiple GPUs, or whether GPU0 is included, it always results in an error (insufficient memory on GPU0), and I am unable to start the training.
